design/0001-frr.md

```diff
@@ -0,0 +1,340 @@
+cluster and not with each other.
+
+## Drawbacks
+
+* This would couple MetalLB with the FRR community so that it would be somewhat
+  dependent on this community for bug-fixes and further feature improvements.
+* As FRR has a larger code base and configuration space, it will have a larger
+  security attack surface relative to that of the current native BGP
+  implementation of MetalLB. This could be mitigated by hardening of the
+  configuration. For example, using [peer filtering](http://docs.frrouting.org/en/latest/bgp.html#peer-filtering)
+  , and [MD5 passwords](http://docs.frrouting.org/en/latest/bgp.html#clicmd-neighborPEERpasswordPASSWORD).
+
+## Alternatives
+
+* In terms of enhancing the MetalLB native BGP implementation, an alternative
+  would be to invest more effort in adding additional features.
+* In terms of enabling another BGP agent, there may be other approaches that
+  could be investigated but from the current investigation, this seems the least
+  intrusive.
+
+### BGP Stack Alternatives
+
+A number of alternative open-source routing stacks were considered (FRR, BIRD,
+GoBGP) as a first target for integration. They were evaluated across a number of
+categories. GoBGP was discounted due to its relatively limited feature set. For
+example, it does not support [BFD](https://github.com/osrg/gobgp/issues/2006).
+FRR and BIRD are well-known and mature stacks which have been deployed in
+production and have active development communities. FRR was selected as the
+first target for integration for the following reasons:
+
+* Integration options: Both BIRD and FRR can be configured by MetalLB using the
+  CLI and/or configuration files. However, FRR also has the option to be
+  configured with a gRPC interface which can generate Go bindings.
+* Extensibility options: FRR provides additional extensibility via its `dplane`
+  and `FPM` interfaces. This may ease integration with Kubernetes networking
+  providers. There does not appear to be equivalents in BIRD. This would
+  enable FRR to integrate with dataplanes other than the Linux kernel. For
+  example, 1) hardware dataplanes 2) an Open Flow configured dataplane like Open
+  vSwitch 3) DPDK dataplane.
+* Features: FRR provides support for some additional multiprotocol extentions
+  such as l2vpn-evpn which may be applicable for some use cases. It also
+  provides support for additional routing protocols which, although not
+  strictly required for the MetalLB use case, may be useful in deployments
+  where a user has modified MetalLB to share the routing stack with other
+  applications (not just MetalLB). Although sharing the routing stack with
+  other applications is an unsupported, non-goal of this enhancement, some users
+  may want to inject custom configurations into the BGP stack. [OSPF](https://github.com/metallb/metallb/issues/256)
+  is one example of an additional protocol requested by the MetalLB community.
```

Comment:
I have done an evaluation and FRR does appear to support more. Some of which may be more relevant/important than others. ISIS and l2vpn are two examples. It's worth reminding that there is nothing in this PR that precludes BIRD for a future enhancement. In fact, we have explicitly stated a goal that this will allow different BGP agents to be plugged in.

Reply:
Ok, can we mention these two that bird does not provide, then? I think that has "historical value" when us (or others) in the future conside other options :)

Reply:
Done